libs/cardano-data/src/Data/Set/NonEmpty.hs

```diff
@@ -0,0 +1,55 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RankNTypes #-}
+
+module Data.Set.NonEmpty (
+  NonEmptySet,
+  fromFoldable,
+  fromSet,
+  singleton,
+  toList,
+  toSet,
+) where
+
+import Cardano.Ledger.Binary (DecCBOR, EncCBOR)
+import Control.DeepSeq (NFData)
+import qualified Data.Foldable as Foldable
+import Data.Set (Set)
+import qualified Data.Set as Set
+import NoThunks.Class (NoThunks)
+import Test.QuickCheck (Arbitrary (arbitrary, shrink))
+import qualified Test.QuickCheck as QC
+
+newtype NonEmptySet a = NonEmptySet (Set a)
+  deriving stock (Show, Eq)
+  deriving newtype (EncCBOR, DecCBOR, NoThunks, NFData)
+
+instance (Arbitrary a, Ord a) => Arbitrary (NonEmptySet a) where
```

Comment:
I see the Arbitrary instances in testlib usually (to avoid a QuickCheck dependency I imagine), but this is here because I needed the constructor and didn't want to add an unsafe constructor. Ideas/preferences on this?